lib/diyepw/diyepw/analyze_noaa_isd_lite_files.py

```diff
@@ -0,0 +1,140 @@
+
+        # Take year-month-day-hour columns and convert to datetime stamps.
+        df['obs_timestamps'] = _pd.to_datetime(_pd.DataFrame({'year': df['Year'],
+                                                            'month': df['Month'],
+                                                            'day': df['Day'],
+                                                            'hour': df['Hour']}))
+
+        # Remove unnecessary year, month, day, hour columns
+        df = df.drop(columns=['Year', 'Month', 'Day', 'Hour'])
+
+        rows_present = df.shape[0]
+        file_description = {
+            'file': file,
+            'total_rows_missing': 8760 - rows_present,
+            'max_consec_rows_missing': 0
+        }
+        # Skip the work of counting consecutive missing rows for files that have no missing rows
+        if file_description['total_rows_missing'] > 0:
+            file_description['max_consec_rows_missing'] = _get_max_missing_rows_from_hourly_dataframe(df, 'obs_timestamps')
+
+        # Depending on how many total and consecutive rows are missing, add the current file to one of our
+        # collections to be returned to the caller
+        if file_description['total_rows_missing'] > max_missing_rows:
+            too_many_missing_rows.append(file_description)
+        elif file_description['max_consec_rows_missing'] > max_consecutive_missing_rows:
+            too_many_consecutive_missing_rows.append(file_description)
+        else:
+            good_files.append(file_description)
+
+    return {
+        "good": good_files,
+        "too_many_total_rows_missing": too_many_missing_rows,
+        "too_many_consecutive_rows_missing": too_many_consecutive_missing_rows
+    }
+
+def _get_max_missing_rows_from_hourly_dataframe(df:_pd.DataFrame, timestamp_col_name:str) ->int:
+    """
+    Given a DataFrame containing hourly timestamps over a year, determine the longest sequence of timestamps
+    missing from that DataFrame.
+
+    :param df:
+    :param timestamp_col_name:
+    :return:
+    """
+    # Create series of continuous timestamp values for that year
+    all_timestamps = _pd.date_range(df[timestamp_col_name].iloc[0], periods=8760, freq='H')
+
+    # Merge to one dataframe containing all continuous timestamp values.
+    all_times = _pd.DataFrame(all_timestamps, columns=['all_timestamps'])
+    df_all_times = _pd.merge(all_times, df, how='left', left_on='all_timestamps', right_on=timestamp_col_name)
+
+    # Create series of only the missing timestamp values
+    missing_times = df_all_times[df_all_times.isnull().any(axis=1)]
+    missing_times = missing_times['all_timestamps']
+
+    if missing_times.empty:
+        return 0
+
+    # Create a series containing the time step distance from the previous timestamp for the missing timestamp values
+    missing_times_diff = missing_times.diff()
+
+    # Count the maximum number of consecutive missing time steps.
+    counter = 1
+    max_missing_rows = 1
```

Comment:
Not that big of a deal here, but you could likely dump this loop with a lag, lead approach in pandas get the diffs.  No need to change this though.

Reply:
I wasn't familiar with Lead/Lag, but it does look like it would work. It works correctly, though, and it's plenty fast relative to the runtime of the application, so I'm inclined to just leave it even if it's not the most elegant solution.